### DISCUSSION TITLE
Way to get current status of a whisper component?

### DISCUSSION THREAD

**Discussion by hans-hsieh-asus** [open-olive/loop-development-kit#139] [Q&A]
Is there a way to get the current status of a whisper component with LDK 3 for javascript? 
Let's say if we want to implement a form-like behavior, we will need to get the status (e.g. content of a text-input) of each whisper component when the submit button is clicked. The `whisper.create()` will resolves a `Whisper` object, but it looks like the object only contains id and label. Or, should we use the callback functions like `onChange` to remember the changes of each component?

**Comment by cshields-oliveai** [comment]
Hello @hans-hsieh-asus 

In LDK version [3.0.0-beta.1](https://github.com/open-olive/loop-development-kit/releases/tag/v3.0.0-beta.1):

You are correct, the function signature for `create` does return a `Whisper` promise: 

`function create(whisper: NewWhisper): Promise<Whisper>`

I did want to point out that the `Whisper` type also includes an `Array<Components>`:

```typescript
interface Whisper {
  id: string;
  label: string;
  components: Array<Components>;
  close(callback: () => void): void;
}
```

From this list of `Components`, it is possible to get all `ChildComponents`. This alone doesn't answer your question, I just wanted to point that out in case it helps your cause.

With the current design of Whispers in the LDK, my current recommended implementation for form-like behavior would be to keep an object, something like `searchOptions` that includes all of the values from each field in your form up to date via the `onChange` callback. Then when your form is "submit" (potentially on a button press event), pass in those `searchOptions`.

If you have any ideas on enhancements for this design, we welcome you to create a Github Issue for the LDK and set the label to "enhancement", or even create a PR for the LDK if are interested.

Thanks again for all of the feedback!

**Reply by hans-hsieh-asus** [reply]
But it looks like the `Whisper` object returned has `components` property being `undefined`.  
Using the `onChange` to record the changes is ok for me. However, it looks like the callbacks wouldn't be called immediately. They are called after a little delay. If the user updates the content of a text field or a checkbox and quickly clicks the submit button, is it possible to get an outdated status of the form?

**Reply by hans-hsieh-asus** [reply]
For the problem of missing `components` property, let me create [an issue](https://github.com/open-olive/loop-development-kit/issues/141) to track this.

**Reply by cshields-oliveai** [reply]
@hans-hsieh-asus I checked into `components` being undefined and you are correct. This is not intended behavior of course so thank you for creating that issue. We will get it resolved as soon as we can.

To address creating your own "form" by tying into `onChange()` - we do realize there are some drawbacks to the way it currently sits, especially when related to timing. I'm happy to report that there is also work planned in the near future to change the way that whispers operate specifically as it relates to forms. I will incorporate this feedback into that work.

As I mentioned in the other thread, we will be deactivating discussions. Feel free to create a feature request for "whisper forms" over in the issues space.
